### ISSUE TITLE
unloadAll does not clear internalModel Record ID map

### ISSUE THREAD

**Issue by cah-briangantzler** [emberjs/data#5041]
Using Ember Data 2.13.1

The backend API I am using required me to specify the id of records, so when a createRecord is given, the id is provided. If a createRecord is added to the store and persisted, then unloadAll is performed on the model, when another record is attempted to be created with the same id, an assert is throwing that the id is already in use. There are no records in the store with that id. It appears that the unloadAll is not clearing an internal record map.

<img width="412" alt="screen shot 2017-06-27 at 1 19 40 pm" src="https://user-images.githubusercontent.com/1529286/27601499-971f1710-5b3d-11e7-9858-b48fdec3b0d4.png">

The assert is in the buildInternalModel method of store. As you can see the id is still recorded in the _idToModel and idToRecord yet the _models array is empty. This is preventing me from adding the same item to another shopping cart after the first cart has been unloaded

**Comment by tylerturdenpants** [comment]
The workaround at https://github.com/travis-ci/travis-web/issues/1173 might be of some help until the issue is addressed.

**Comment by cah-briangantzler** [comment]
Thanks, that does appear what needs to be done. I am doing an unloadAll, can adapt the code. Hoping that makes it into Ember-data proper sometime.

**Comment by acorncom** [comment]
Related issue: https://github.com/emberjs/data/issues/4972

**Comment by lolmaus** [comment]
@tylerturdenpants Thank you. :bow: This fix worked for us.

I must add that a simple `record.unloadRecord()` fails for us (Ember Data 2.15.3). ED seems to enter an inconsistent state. The actual error is `cannot call eachAttribute on null` somewhere deep in ED internals.

A workaround I found is to destroy the record before unloading it:

```js
record
  .destroyRecord()
  .then(() => {
    record.unloadRecord()
    // the Travis fix goes here
  })
```

**Comment by lolmaus** [comment]
This is the specific problem I'm talking about: https://github.com/emberjs/data/issues/4963

**Comment by tylerturdenpants** [comment]
@lolmaus you are welcome. It's been a huge challenge to move from 2.12 because my app, which loads and unloads very frequently (I deal with quasi real time pricing). The is the first time in 2 years I've been this far behind in a single core package of Ember. 

It deeply saddens me that the plethora of unload issues still exist across 3 published versions of Ember, namely 2.13, 2.14, and 2.15. While I can't contribute to the solution by submitting PR's, I can provide some financial incentives out of my own pocket. Hopefully this issue will get a core champion soon. 

Since there are so any of us affected by these issues, maybe we could come together to offer core members something? to pause on features scheduled for 2.16 and finally put an end to this ongoing unload nightmare. 

I hope this reply is taken in the right way. I love Ember. I lover Ember data. But I simply cannot give back. The Ember data internals are way above my skill set. I understand the value of open source and the core members who give their free time.

**Comment by efx** [comment]
Commented on the other issue, but it bears linking here: https://github.com/emberjs/data/issues/4972#issuecomment-351166147

**Comment by runspired** [comment]
I believe this was resolved as of #5378, if not please reopen
